src/Grpc.NetCore.HttpClient/Internal/GrpcCall.cs

```diff
@@ -0,0 +1,433 @@
+        {
+            // Validate deadline before creating any objects that require cleanup
+            ValidateDeadline(options.Deadline);
+
+            _callCts = new CancellationTokenSource();
+            Method = method;
+            Options = options;
+            _clock = clock;
+
+            if (options.CancellationToken.CanBeCanceled)
+            {
+                // The cancellation token will cancel the call CTS
+                _ctsRegistration = options.CancellationToken.Register(CancelCall);
+            }
+
+            if (options.Deadline != null && options.Deadline != DateTime.MaxValue)
+            {
+                var timeout = options.Deadline.Value - _clock.UtcNow;
+                _timeout = (timeout > TimeSpan.Zero) ? timeout : TimeSpan.Zero;
+            }
+
+            if (_timeout != null)
+            {
+                // Deadline timer will cancel the call CTS
+                _deadlineTimer = new Timer(DeadlineExceeded, null, _timeout.Value, Timeout.InfiniteTimeSpan);
+            }
+        }
+
+        private void ValidateDeadline(DateTime? deadline)
+        {
+            if (deadline != null && deadline != DateTime.MaxValue && deadline != DateTime.MinValue && deadline.Value.Kind != DateTimeKind.Utc)
+            {
+                throw new InvalidOperationException("Deadline must have a kind DateTimeKind.Utc or be equal to DateTime.MaxValue or DateTime.MinValue.");
+            }
+        }
+
+        public CancellationToken CancellationToken
+        {
+            get { return _callCts.Token; }
+        }
+
+        public bool IsCancellationRequested
+        {
+            get { return _callCts.IsCancellationRequested; }
+        }
+
+        public void StartUnary(System.Net.Http.HttpClient client, TRequest request)
+        {
+            var message = CreateHttpRequestMessage();
+            SetMessageContent(request, message);
+            StartSend(client, message);
+        }
+
+        public void StartClientStreaming(System.Net.Http.HttpClient client)
+        {
+            var message = CreateHttpRequestMessage();
+            ClientStreamWriter = CreateWriter(message);
+            StartSend(client, message);
+        }
+
+        public void StartServerStreaming(System.Net.Http.HttpClient client, TRequest request)
+        {
+            var message = CreateHttpRequestMessage();
+            SetMessageContent(request, message);
+            StartSend(client, message);
+            ClientStreamReader = new HttpContentClientStreamReader<TRequest, TResponse>(this);
+        }
+
+        public void StartDuplexStreaming(System.Net.Http.HttpClient client)
+        {
+            var message = CreateHttpRequestMessage();
+            ClientStreamWriter = CreateWriter(message);
+            StartSend(client, message);
+            ClientStreamReader = new HttpContentClientStreamReader<TRequest, TResponse>(this);
+        }
+
+        public void Dispose()
```

Comment:
Is there a reason why the type isn't disposable?

Reply:
It was never be used as IDisposable (cast as IDisposable, or used with a using statement). On the other hand it doesn't hurt so I can change it.